pennylane/devices/default_qubit.py

```diff
@@ -887,3 +887,66 @@ def analytic_probability(self, wires=None):
         real_state = self._real(flat_state)
         imag_state = self._imag(flat_state)
         return self.marginal_prob(real_state**2 + imag_state**2, wires)
+
+    def classical_shadow(self, wires, n_snapshots, circuit, seed=None):
+        """TODO: docs"""
+
+        n_qubits = len(self.wires)
+        device_wires = np.array(self.map_wires(wires))
+
+        if seed is not None:
+            rng = np.random.RandomState(seed)
+            recipes = rng.randint(0, 3, size=(n_snapshots, n_qubits))
+        else:
+            recipes = np.random.randint(0, 3, size=(n_snapshots, n_qubits))
+
+        obs_list = self._stack(
+            [
+                qml.PauliX.compute_matrix(),
+                qml.PauliY.compute_matrix(),
+                qml.PauliZ.compute_matrix(),
+            ]
+        )
+        uni_list = self._stack(
+            [
+                qml.Hadamard.compute_matrix(),
+                qml.Hadamard.compute_matrix() @ qml.RZ.compute_matrix(-np.pi / 2),
+                qml.Identity.compute_matrix(),
+            ]
+        )
+        obs = obs_list[recipes]
+        uni = uni_list[recipes]
+
+        outcomes = np.zeros((n_snapshots, n_qubits))
+        stacked_state = self._stack([self._state for _ in range(n_snapshots)])
+
+        for i in range(n_qubits):
+
+            # trace out every qubit except the first
+            first_qubit_state = self._einsum(
+                f"{ABC[n_qubits - i + 1]}{ABC[:n_qubits - i]},{ABC[n_qubits - i + 1]}{ABC[n_qubits - i]}{ABC[1:n_qubits - i]}"
+                f"->{ABC[n_qubits - i + 1]}a{ABC[n_qubits - i]}",
+                stacked_state,
+                self._conj(stacked_state),
+            )
+
+            # sample the observables on the first qubit
+            probs = (self._einsum("abc,acb->a", first_qubit_state, obs[:, i]) + 1) / 2
+            samples = np.random.uniform(0, 1, size=probs.shape) > probs
+            outcomes[:, i] = samples
+
+            # collapse the state
+            rotated_state = self._einsum("ab...,acb->ac...", stacked_state, uni[:, i])
+            stacked_state = rotated_state[np.arange(n_snapshots), self._cast(samples, np.uint8)]
```

Comment:
I am not sure I understand this line. What is the reason for taking the `samples >probs` of the second index of `rotated_state` and how is that possible in the first place? `samples` is of size `(n_snapshots)` and `stacked_state.shape = (n_snapshots, 2, 2, 2,..)`?

Reply:
The second index of `rotated_state` is just the sampled bits (which we cast to ints). If 0 is sampled, then we take the first row corresponding to the collapsed state, which corresponds to the post-measurement state if 0 is measured in the first qubit. Similarly if 1 is sampled, we take the second row.

Reply:
Marking as resolved.